fenix/app/src/main/java/org/mozilla/fenix/share/SaveToPDFMiddleware.kt

```diff
@@ -31,15 +41,176 @@ class SaveToPDFMiddleware(
+            is EngineAction.SaveToPdfAction -> {
+                postTelemetryTapped(ctx.state.findTab(action.tabId))
+                // Continue to generate the PDF, passing through here to add telemetry
+                next(action)
+            }
+
+            is EngineAction.SaveToPdfCompleteAction -> {
+                postTelemetryCompleted(ctx.state.findTab(action.tabId))
+            }
+
+            is EngineAction.SaveToPdfExceptionAction -> {
+                // See https://github.com/mozilla-mobile/fenix/issues/27649 for more details,
+                // why a Toast is used here.
+                ThreadUtils.runOnUiThread {
+                    Toast.makeText(context, R.string.unable_to_save_to_pdf_error, LENGTH_LONG).show()
+                }
+
+                postTelemetryFailed(ctx.state.findTab(action.tabId), action.throwable)
+            }
+            else -> {
+                next(action)
+            }
+        }
+    }
+
+    /**
+     * Use to generate failure extra reasons for Save To PDF failure telemetry.
+     *
+     * @param exception - A given exception that will be properly labeled for telemetry posting.
+     * @return processed failure reason to send in telemetry.
+     */
+    /* package */ @VisibleForTesting
+    fun telemetryErrorReason(exception: Exception): String {
+        var failureMsg = "unknown"
+        if (exception is GeckoSession.GeckoPrintException) {
+            when (exception.code) {
+                ERROR_PRINT_SETTINGS_SERVICE_NOT_AVAILABLE -> failureMsg = "no_settings_service"
+                ERROR_UNABLE_TO_CREATE_PRINT_SETTINGS -> failureMsg = "no_settings"
+                ERROR_UNABLE_TO_RETRIEVE_CANONICAL_BROWSING_CONTEXT -> failureMsg = "no_canonical_context"
+                ERROR_NO_ACTIVITY_CONTEXT_DELEGATE -> failureMsg = "no_activity_context_delegate"
+                ERROR_NO_ACTIVITY_CONTEXT -> failureMsg = "no_activity_context"
+            }
+        }
+        if (exception is IOException) {
+            failureMsg = "io_error"
+        }
+        return failureMsg
+    }
+
+    /**
+     * Use to generate extra sources for Save To PDF telemetry.
+     *
+     * @param isPdfJs - If the page is a PDF.js page or not.
+     * @return processed page source type to send in telemetry.
+     */
+    /* package */ @VisibleForTesting
+    fun telemetrySource(isPdfJs: Boolean?): String {
+        val source = when (isPdfJs) {
+            null -> "unknown"
+            true -> "pdf"
+            false -> "non-pdf"
+        }
+        return source
+    }
+
+    /**
+     * Indicates the Save As PDF action was requested and posts telemetry via Glean.
+     *
+     * @param tab - tab state to use for page source category
+     */
+    private fun postTelemetryTapped(tab: TabSessionState?) {
+        ThreadUtils.postToBackgroundThread() {
+            if (tab != null && tab.engineState.engineSession != null) {
+                tab.engineState.engineSession!!.checkForPdfJs(
```

Comment:
I made an update, but I'm not sure if I updated it correctly. I think I might have the `SaveAsPdfMiddlewareTest` setup incorrectly because it requires the `else` path.

On `SaveAsPdfMiddlewareTest`, I had trouble setting up a mock `GeckoEngineSession` that can respond to `isPdfJs` similarly to how `GeckoEngineSessionTest` mocks. When setting up similarly and adding    `whenever(geckoSession.isPdfJs).thenReturn(GeckoResult.fromValue(true))` in the middleware I get a `Cannot access class 'org.mockito.stubbing.OngoingStubbing`. Not sure where to take it, any tips?

Reply:
Thanks @jonalmeida and @Amejia481 for the help on Slack getting the telemetry test configured properly!